backend/src/main/java/ai/verta/modeldb/batchProcess/CollaboratorResourceMigration.java

```diff
@@ -0,0 +1,236 @@
+        if (ModelDBUtils.needToRetry(ex)) {
+          migrateProjects();
+        } else {
+          throw ex;
+        }
+      }
+    }
+
+    LOGGER.debug("Projects migration finished");
+  }
+
+  private static Long getEntityCount(Class<?> klass) {
+    try (Session session = ModelDBHibernateUtil.getSessionFactory().openSession()) {
+      CriteriaBuilder criteriaBuilder = session.getCriteriaBuilder();
+      CriteriaQuery<Long> countQuery = criteriaBuilder.createQuery(Long.class);
+      countQuery.select(criteriaBuilder.count(countQuery.from(klass)));
+      return session.createQuery(countQuery).getSingleResult();
+    } catch (Exception ex) {
+      if (ModelDBUtils.needToRetry(ex)) {
+        return getEntityCount(klass);
+      } else {
+        throw ex;
+      }
+    }
+  }
+
+  private void createWorkspacePermissions(
+      Optional<String> workspaceName,
+      Optional<WorkspaceTypeEnum.WorkspaceType> workspaceType,
+      String resourceId,
+      Optional<Long> ownerId,
+      ModelDBServiceResourceTypes resourceType,
+      CollaboratorTypeEnum.CollaboratorType collaboratorType,
+      VisibilityEnum.Visibility visibility) {
+    try (AuthServiceChannel authServiceChannel = new AuthServiceChannel()) {
+      LOGGER.info("Calling CollaboratorService to create resources");
+      ResourceType modeldbServiceResourceType =
+          ResourceType.newBuilder().setModeldbServiceResourceType(resourceType).build();
+      Resources resources =
+          Resources.newBuilder()
+              .setResourceType(modeldbServiceResourceType)
+              .setService(ServiceEnum.Service.MODELDB_SERVICE)
+              .addResourceIds(resourceId)
+              .build();
+      ResourceVisibility resourceVisibility = getResourceVisibility(workspaceType, visibility);
+      SetResources.Builder setResourcesBuilder =
+          SetResources.newBuilder().setResources(resources).setVisibility(resourceVisibility);
+
+      if (resourceVisibility.equals(ResourceVisibility.ORG_DEFAULT)) {
+        setResourcesBuilder.setCollaboratorType(collaboratorType);
+      }
+
+      if (ownerId.isPresent()) {
+        setResourcesBuilder.setOwnerId(ownerId.get());
+      }
+      if (workspaceName.isPresent()) {
+        setResourcesBuilder = setResourcesBuilder.setWorkspaceName(workspaceName.get());
+      } else {
+        throw new IllegalArgumentException(
+            "workspaceId and workspaceName are both empty.  One must be provided.");
+      }
+      SetResources setResources = setResourcesBuilder.build();
+      SetResources.Response setResourcesResponse =
+          authServiceChannel.getCollaboratorServiceBlockingStub().setResources(setResources);
+
+      LOGGER.info("SetResources message sent.  Response: " + setResourcesResponse);
+    } catch (StatusRuntimeException ex) {
+      LOGGER.error(ex);
+      ModelDBUtils.retryOrThrowException(ex, false, retry -> null);
+    }
+  }
+
+  private ResourceVisibility getResourceVisibility(
```

Comment:
Not used if we delete `createWorkspacePermissions`.

Reply:
We can not remove this https://github.com/VertaAI/modeldb/pull/1753/files/fe66220573e9ebb12c2781b45257a825db853f0a#r542103742

Reply:
Let's resolve the duplicated code and see if this is needed